ykcapi/src/lib.rs

```diff
@@ -147,20 +147,70 @@ pub struct LiveAOTVals {
     length: usize,
 }
 
-/// Reads the stackmap and saved registers from the given address (i.e. the return address of the
-/// deoptimisation call).
+/// After a guard failure, reconstructs the stack frames and registers and then jumps back to the
+/// AOT code from where to continue.
+#[cfg(target_arch = "x86_64")]
+#[naked]
+#[no_mangle]
+pub extern "C" fn __ykrt_reconstruct_frames(
+    newframesptr: *const c_void,
+) {
+    unsafe {
+        asm!(
+            // The first 8 bytes of the new frames is the size of the map, needed for copying it
+            // over. Move it into rdx.
+            "mov rdx, [rdi]",
+            // Then adjust the address to where the new stack actually starts.
+            "add rdi, 8",
+            // Make space for the new stack, but use 8 bytes less in order to overwrite this
+            // function's return address since we won't be returning there.
+            "add rsp, 8",
+            "sub rsp, rdx",
+            // Copy over the new stack frames.
+            "mov rsi, rdi", // 2nd arg: src
+            "mov rdi, rsp", // 1st arg: dest
+            "call memcpy",
+            // Now move the source (i.e. the heap allocated frames) into the first argument and its
+            // size into the second. Then free the memory.
+            "mov rdi, rsi",
+            "mov rsi, rdx",
+            "call munmap",
+            // Restore registers.
+            "pop r15",
+            "pop r14",
+            "pop r13",
+            "pop r12",
+            "pop r11",
+            "pop r10",
+            "pop r9",
+            "pop r8",
+            "pop rbp",
+            "pop rdi",
+            "pop rsi",
+            "pop rbx",
+            "pop rcx",
+            "pop rdx",
+            "pop rax",
```

Comment:
Can we put a FIXME in here to remind us to restore floating point registers at some stage?

Reply:
Fixed in [7660a3c](https://github.com/ykjit/yk/pull/578/commits/7660a3c9813795998bfb6a958058c2d30b60fc9f)